### ISSUE TITLE
IDE0066 Suggests code that doesn't compile for implicitly convertable types

### ISSUE THREAD

**Issue by xdaDaveShaw** [dotnet/roslyn#44572]
**Version Used**: Visual Studio 16.5.4 (also VS Code, latest stable)

**Steps to Reproduce**:

1. Given a class `C` like so:
```c#
class C
{
	public C(String s) => _s = s;
	private readonly String _s;
	public static implicit operator String(C value) => value._s;
	public static implicit operator C(String value) => new C(value);
}
```
2. Write a method with a switch statement on and instance of `C`:
```c#
switch (c)
{
	case "A": return true;
	default: return false;
}
``` 
3. IDE0066 suggests a switch expression as follows:
```c#
return c switch
{
	"A" => true,
	_ => false,
};
```
This does not compile because:

> CS0150	A constant value is expected

**Expected Behavior**:
Either IDE0066 doesn't trigger where the generated code won't compile

or

Switch expressions can work the same as switch statements with regards to implicit conversion from constants.

**Actual Behavior**:
As above.

**Comment by CyrusNajmabadi** [comment]
This is actually a bit surprising.  @gafter could you explain how the language sees things differently for the switch expression versus teh switch statement here?

**Comment by gafter** [comment]
A switch statement on a type that has a conversion to the type `string` applies that conversion before doing the switch. That is a legacy behavior of C# switch statements. But it doesn't occur for the switch expression. So the switch expression's input type is `C`. The constant pattern `"A"` is converted to the input type, in this case using the conversion operator, and results in a value that is not a constant. Since the pattern is required to be a constant, that is an error.

**Comment by CyrusNajmabadi** [comment]
Got it.  We'll have to check for that concept in our 'convert to switch' logic.  @alrz do you want to take this?  or would you prefer we do it?

**Comment by alrz** [comment]
I'm in the process moving out right now, I don't think I can get to it any time soon. :)

**Comment by CyrusNajmabadi** [comment]
We would take a targeted community fix for this.

**Comment by nalka0** [comment]
Is this still meant to be open?

Likely since #45461, it now generates:
```cs
return (string)c switch
{
    "A" => true,
    _ => false,
};
```
which doesn't error.
